app/assets/javascripts/discourse/app/lib/sticky-avatars.js

```diff
@@ -0,0 +1,101 @@
+import Site from "discourse/models/site";
+import { bind } from "discourse-common/utils/decorators";
+import { schedule } from "@ember/runloop";
+
+export default class StickyAvatars {
+  stickyClass = "sticky-avatar";
+  topicPostSelector = "#topic .post-stream .topic-post";
+  intersectionObserver = null;
+  direction = "⬇️";
+  prevOffset = -1;
+
+  static init(api) {
+    new this(api).init();
+  }
+
+  constructor(api) {
+    this.api = api;
+  }
+
+  init() {
+    if (Site.currentProp("mobileView")) {
+      return;
+    }
+
+    this.api.onAppEvent("topic:current-post-scrolled", this._handlePostNodes);
+    this.api.onAppEvent("topic:scrolled", this._handleScroll);
+    this.api.onAppEvent("page:topic-loaded", this._initIntersectionObserver);
+    this.api.cleanupStream(this._clearIntersectionObserver);
+  }
+
+  @bind
+  _handleScroll(offset) {
+    if (offset <= 0) {
+      this.direction = "⬇️";
+      document
+        .querySelectorAll(`${this.topicPostSelector}.${this.stickyClass}`)
+        .forEach((node) => node.classList.remove(this.stickyClass));
+    } else if (offset > this.prevOffset) {
+      this.direction = "⬇️";
+    } else {
+      this.direction = "⬆️";
+    }
+    this.prevOffset = offset;
+  }
+
+  @bind
+  _handlePostNodes() {
+    this._clearIntersectionObserver();
+
+    schedule("afterRender", () => {
+      this._initIntersectionObserver();
+
+      document.querySelectorAll(this.topicPostSelector).forEach((postNode) => {
+        this.intersectionObserver.observe(postNode);
+
+        const topicAvatarNode = postNode.querySelector(".topic-avatar");
+        if (!topicAvatarNode || !postNode.querySelector("#post_1")) {
+          return;
+        }
+
+        const topicMapNode = postNode.querySelector(".topic-map");
+        if (!topicMapNode) {
+          return;
+        }
+        topicAvatarNode.style.marginBottom = `${topicMapNode.clientHeight}px`;
+      });
+    });
+  }
+
+  @bind
+  _initIntersectionObserver() {
+    const headerHeight = document.querySelector(".d-header")?.clientHeight || 0;
+
+    this.intersectionObserver = new IntersectionObserver(
+      (entries) => {
+        entries.forEach((entry) => {
+          if (!entry.isIntersecting || entry.intersectionRatio === 1) {
+            entry.target.classList.remove(this.stickyClass);
+            return;
+          }
+
+          const postContentHeight = entry.target.querySelector(".contents")
+            ?.clientHeight;
+          if (
+            this.direction === "⬆️" ||
+            postContentHeight > window.innerHeight - headerHeight
```

Comment:
This is extremely minor, but after living with this for a bit, I wonder whether it would be better to remove the first part of this OR clause, `this.direction === "⬆️"`, that way whether you are scrolling up or down, the same avatars would be sticky.

Reply:
It was done like that originally actually :slightly_smiling_face: Only long enough posts were sticky but when scrolling up it’s not always clear who’s the author (even with small posts) so we decided to make all posts sticky when scrolling up.